contracts/core/Core.sol

```diff
@@ -219,4 +234,87 @@ contract Core is
             _quantity
         );
     }
+
+    /**
+     * Function to convert Set Tokens into underlying components
+     *
+     * @param _tokenAddress    The address of the Set token
+     * @param _quantity        The number of tokens to redeem
+     */
+    function redeem(
+        address _tokenAddress,
+        uint _quantity
+    )
+        public
+        isValidTokenAddress(_tokenAddress)
+        isNonZero(_quantity)
+    {
+        // Check if multiple of naturalUnit
+        uint naturalUnit = SetToken(_tokenAddress).naturalUnit();
+        require(isMultipleOfNaturalUnit(_quantity, naturalUnit));
+
+        // Burn the Set token
+        SetToken(_tokenAddress).burn(_quantity);
+
+        // Decrement the Set Token amount (this will validate if balance has enough)
+        Vault(vaultAddress).decrementTokenOwner(
+            msg.sender,
+            _tokenAddress,
+            _quantity
+        );
+
+        // Transfer the underlying tokens to the corresponding token balances
+        address[] memory components = SetToken(_tokenAddress).getComponents();
+        uint[] memory units = SetToken(_tokenAddress).getUnits();
+        for (uint16 i = 0; i < components.length; i++) {
+            address currentComponent = components[i];
+            uint currentUnits = units[i];
+
+            uint preTransferBalance = ERC20(currentComponent).balanceOf(msg.sender);
+
+            uint transferValue = calculateTransferValue(
+                currentUnits,
+                naturalUnit,
+                _quantity
+            );
+            require(ERC20(currentComponent).transfer(msg.sender, transferValue));
```

Comment:
There shouldn't be any transferring in redeem. Just decrementing / incrementing balances of the vault

Reply:
@felix2feng do we have any incrementing of the vault here? if we're redeeming, its just decrement right?